core/src/avm2/activation.rs

```diff
@@ -815,23 +815,39 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
         };
 
         if let Some(body) = method.body() {
+            // Use `coerce_to_object` so that we handle primitives correctly.
+            let err_object = error.coerce_to_object(self);
             for e in body.exceptions.iter() {
                 if instruction_start >= e.from_offset as usize
                     && instruction_start < e.to_offset as usize
-                    && e.type_name.0 == 0
-                // Currently we support only typeless catch clauses
                 {
-                    // Emulate pushing the exception object
-                    self.context.avm2.push(error);
+                    let mut matches = false;
+                    // A typeless catch block (e.g. `catch(er) { ... }`) will
+                    // always match.
+                    if e.type_name.0 == 0 {
+                        matches = true;
+                    } else if let Ok(err_object) = err_object {
+                        let type_name = self.pool_multiname_static(method, e.type_name)?;
+                        let ty_class = self.resolve_class(&type_name)?;
+
+                        matches = err_object.is_of_type(ty_class, self);
```

Comment:
`is_of_type` is also on `Value`, so you don't need to call `coerce_to_object` yourself.
I think then you could change the branching to `let matches = if (...) { true } else { ... is_of_type() }`.

Reply:
This is a little more efficient, since we'll only call `coerce_to_object` once.